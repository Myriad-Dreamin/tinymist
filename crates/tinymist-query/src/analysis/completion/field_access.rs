//! Completion for field access on nodes.

use typst::syntax::ast::MathTextKind;

use crate::analysis::completion::typst_specific::ValueCompletionInfo;

use super::*;
impl CompletionPair<'_, '_, '_> {
    /// Add completions for all dot targets on a node.
    pub fn doc_access_completions(&mut self, target: &LinkedNode) -> Option<()> {
        self.value_dot_access_completions(target)
            .or_else(|| self.type_dot_access_completions(target))
    }

    /// Add completions for all fields on a type.
    fn type_dot_access_completions(&mut self, target: &LinkedNode) -> Option<()> {
        let mode = self.cursor.leaf_mode();

        if matches!(mode, InterpretMode::Math) {
            return None;
        }

        self.type_field_access_completions(target);
        Some(())
    }

    /// Add completions for all fields on a type.
    fn type_field_access_completions(&mut self, target: &LinkedNode) -> Option<()> {
        let ty = self
            .worker
            .ctx
            .post_type_of_node(target.clone())
            .filter(|ty| !matches!(ty, Ty::Any));
        crate::log_debug_ct!("type_field_access_completions_on: {target:?} -> {ty:?}");
        let mut defines = Defines {
            types: self.worker.ctx.type_check(&self.cursor.source),
            defines: Default::default(),
            docs: Default::default(),
        };
        ty?.iface_surface(
            true,
            &mut CompletionScopeChecker {
                check_kind: ScopeCheckKind::FieldAccess,
                defines: &mut defines,
                ctx: self.worker.ctx,
            },
        );

        self.def_completions(defines, true);
        Some(())
    }

    /// Add completions for all fields on a value.
    fn value_dot_access_completions(&mut self, target: &LinkedNode) -> Option<()> {
        let (value, styles) = self.worker.ctx.analyze_expr(target).into_iter().next()?;

        let mode = self.cursor.leaf_mode();
        let valid_field_access_syntax =
            !matches!(mode, InterpretMode::Math) || is_valid_math_field_access(target);
        let valid_postfix_target =
            !matches!(mode, InterpretMode::Math) || is_valid_math_postfix(target);

        if !valid_field_access_syntax && !valid_postfix_target {
            return None;
        }

        if valid_field_access_syntax {
            self.value_field_access_completions(&value, mode);
        }
        if valid_postfix_target {
            self.postfix_completions(target, Ty::Value(InsTy::new(value.clone())));
        }

        match value {
            Value::Symbol(symbol) => {
                self.symbol_var_completions(&symbol, None);

                if valid_postfix_target {
                    self.ufcs_completions(target);
                }
            }
            Value::Content(content) => {
                if valid_field_access_syntax {
                    for (name, value) in content.fields() {
                        self.value_completion(Some(name.into()), &value, false, None);
                    }
                }
                if valid_postfix_target {
                    self.ufcs_completions(target);
                }
            }
            Value::Dict(dict) if valid_field_access_syntax => {
                for (name, value) in dict.iter() {
                    self.value_completion(Some(name.clone().into()), value, false, None);
                }
            }
            Value::Func(func) if valid_field_access_syntax => {
                // Autocomplete get rules.
                if let Some((elem, styles)) = func.element().zip(styles.as_ref()) {
                    for param in elem.params().iter().filter(|param| !param.required) {
                        if let Some(value) = elem
                            .field_id(param.name)
                            .map(|id| elem.field_from_styles(id, StyleChain::new(styles)))
                        {
                            self.value_completion(
                                Some(param.name.into()),
                                &value.unwrap(),
                                false,
                                None,
                            );
                        }
                    }
                }
            }
            _ => {}
        }

        Some(())
    }

    fn value_field_access_completions(&mut self, value: &Value, mode: InterpretMode) {
        let elem_parens = !matches!(mode, InterpretMode::Math);
        for (name, bind) in value.ty().scope().iter() {
            if matches!(mode, InterpretMode::Math) && is_func(bind.read()) {
                continue;
            }

            self.value_completion_(
                bind.read(),
                ValueCompletionInfo {
                    label: Some(name.clone()),
                    parens: elem_parens,
                    docs: None,
                    label_details: None,
                    bound_self: true,
                },
            );
        }

        if let Some(scope) = value.scope() {
            for (name, bind) in scope.iter() {
                if matches!(mode, InterpretMode::Math) && is_func(bind.read()) {
                    continue;
                }

                self.value_completion_(
                    bind.read(),
                    ValueCompletionInfo {
                        label: Some(name.clone()),
                        parens: elem_parens,
                        docs: None,
                        label_details: None,
                        bound_self: false,
                    },
                );
            }
        }

        for &field in fields_on(value.ty()) {
            // Complete the field name along with its value. Notes:
            // 1. No parentheses since function fields cannot currently be called
            // with method syntax;
            // 2. We can unwrap the field's value since it's a field belonging to
            // this value's type, so accessing it should not fail.
            self.value_completion_(
                &value.field(field, ()).unwrap(),
                ValueCompletionInfo {
                    label: Some(field.into()),
                    parens: false,
                    docs: None,
                    label_details: None,
                    bound_self: true,
                },
            );
        }
    }
}

fn is_func(read: &Value) -> bool {
    matches!(read, Value::Func(func) if func.element().is_none())
}

fn is_valid_math_field_access(target: &SyntaxNode) -> bool {
    if let Some(field_access) = target.cast::<ast::FieldAccess>() {
        return is_valid_math_field_access(field_access.target().to_untyped());
    }
    if matches!(target.kind(), SyntaxKind::Ident | SyntaxKind::MathIdent) {
        return true;
    }

    false
}

fn is_valid_math_postfix(target: &SyntaxNode) -> bool {
    fn bad_punc_text(punc: char) -> bool {
        punc.is_ascii_punctuation() || punc.is_ascii_whitespace()
    }

    if let Some(target) = target.cast::<ast::MathText>() {
        return match target.get() {
            MathTextKind::Character(ch) => !bad_punc_text(ch),
            MathTextKind::Number(..) => true,
        };
    }

    if let Some(target) = target.cast::<ast::Text>() {
        let target = target.get();
        return !target.is_empty() && target.chars().all(|ch| !bad_punc_text(ch));
    }

    true
}
