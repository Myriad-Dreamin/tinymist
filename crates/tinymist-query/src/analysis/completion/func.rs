//! Completion for functions on nodes.

use super::*;

impl CompletionPair<'_, '_, '_> {
    pub fn func_completion(
        &mut self,
        mode: InterpretMode,
        fn_feat: FnCompletionFeat,
        name: EcoString,
        label_details: Option<EcoString>,
        detail: Option<EcoString>,
        parens: bool,
    ) {
        let base = Completion {
            kind: CompletionKind::Func,
            label_details,
            detail,
            command: self
                .worker
                .ctx
                .analysis
                .trigger_on_snippet_with_param_hint(true)
                .map(From::from),
            ..Default::default()
        };

        if matches!(
            self.cursor.surrounding_syntax,
            SurroundingSyntax::ShowTransform
        ) && (fn_feat.min_pos() > 0 || fn_feat.min_named() > 0)
        {
            self.push_completion(Completion {
                label: eco_format!("{name}.with"),
                apply: Some(eco_format!("{name}.with(${{}})")),
                ..base.clone()
            });
        }
        if fn_feat.is_element
            && matches!(self.cursor.surrounding_syntax, SurroundingSyntax::Selector)
        {
            self.push_completion(Completion {
                label: eco_format!("{name}.where"),
                apply: Some(eco_format!("{name}.where(${{}})")),
                ..base.clone()
            });
        }

        let bad_instantiate = matches!(
            self.cursor.surrounding_syntax,
            SurroundingSyntax::Selector | SurroundingSyntax::SetRule
        ) && !fn_feat.is_element;
        if !bad_instantiate {
            if !parens || matches!(self.cursor.surrounding_syntax, SurroundingSyntax::Selector) {
                self.push_completion(Completion {
                    label: name,
                    ..base
                });
            } else if (fn_feat.min_pos() < 1 || fn_feat.has_only_self()) && !fn_feat.has_rest {
                self.push_completion(Completion {
                    apply: Some(eco_format!("{}()${{}}", name)),
                    label: name,
                    ..base
                });
            } else {
                let accept_content_arg = fn_feat.next_arg_is_content && !fn_feat.has_rest;
                let scope_reject_content = matches!(mode, InterpretMode::Math)
                    || matches!(
                        self.cursor.surrounding_syntax,
                        SurroundingSyntax::Selector | SurroundingSyntax::SetRule
                    );
                self.push_completion(Completion {
                    apply: Some(eco_format!("{name}(${{}})")),
                    label: name.clone(),
                    ..base.clone()
                });
                if !scope_reject_content && accept_content_arg {
                    self.push_completion(Completion {
                        apply: Some(eco_format!("{name}[${{}}]")),
                        label: eco_format!("{name}.bracket"),
                        ..base
                    });
                };
            }
        }
    }
}
