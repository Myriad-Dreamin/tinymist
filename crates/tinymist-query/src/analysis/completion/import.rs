//! Completion for import items.

use super::*;
impl CompletionPair<'_, '_, '_> {
    /// Complete imports.
    pub fn complete_imports(&mut self) -> bool {
        // On the colon marker of an import list:
        // "#import "path.typ":|"
        if_chain! {
            if matches!(self.cursor.leaf.kind(), SyntaxKind::Colon);
            if let Some(parent) = self.cursor.leaf.clone().parent();
            if let Some(ast::Expr::Import(import)) = parent.get().cast();
            if !matches!(import.imports(), Some(ast::Imports::Wildcard));
            if let Some(source) = parent.children().find(|child| child.is::<ast::Expr>());
            then {
                let items = match import.imports() {
                    Some(ast::Imports::Items(items)) => items,
                    _ => Default::default(),
                };

                self.cursor.from = self.cursor.cursor;

                self.import_item_completions(items, vec![], &source);
                if items.iter().next().is_some() {
                    self.worker.enrich("", ", ");
                }
                return true;
            }
        }

        // Behind an import list:
        // "#import "path.typ": |",
        // "#import "path.typ": a, b, |".
        if_chain! {
            if let Some(prev) = self.cursor.leaf.prev_sibling();
            if let Some(ast::Expr::Import(import)) = prev.get().cast();
            if !self.cursor.text[prev.offset()..self.cursor.cursor].contains('\n');
            if let Some(ast::Imports::Items(items)) = import.imports();
            if let Some(source) = prev.children().find(|child| child.is::<ast::Expr>());
            then {
                self.  cursor.from = self.cursor.cursor;
                self.import_item_completions(items, vec![], &source);
                return true;
            }
        }

        // Behind a comma in an import list:
        // "#import "path.typ": this,|".
        if_chain! {
            if matches!(self.cursor.leaf.kind(), SyntaxKind::Comma);
            if let Some(parent) = self.cursor.leaf.clone().parent();
            if parent.kind() == SyntaxKind::ImportItems;
            if let Some(grand) = parent.parent();
            if let Some(ast::Expr::Import(import)) = grand.get().cast();
            if let Some(ast::Imports::Items(items)) = import.imports();
            if let Some(source) = grand.children().find(|child| child.is::<ast::Expr>());
            then {
                self.import_item_completions(items, vec![], &source);
                self.worker.enrich(" ", "");
                return true;
            }
        }

        // Behind a half-started identifier in an import list:
        // "#import "path.typ": th|".
        if_chain! {
            if matches!(self.cursor.leaf.kind(), SyntaxKind::Ident | SyntaxKind::Dot);
            if let Some(path_ctx) = self.cursor.leaf.clone().parent();
            if path_ctx.kind() == SyntaxKind::ImportItemPath;
            if let Some(parent) = path_ctx.parent();
            if parent.kind() == SyntaxKind::ImportItems;
            if let Some(grand) = parent.parent();
            if let Some(ast::Expr::Import(import)) = grand.get().cast();
            if let Some(ast::Imports::Items(items)) = import.imports();
            if let Some(source) = grand.children().find(|child| child.is::<ast::Expr>());
            then {
                if self.cursor.leaf.kind() == SyntaxKind::Ident {
                    self.cursor.from = self.cursor.leaf.offset();
                }
                let path = path_ctx.cast::<ast::ImportItemPath>().map(|path| path.iter().take_while(|ident| ident.span() != self.cursor.leaf.span()).collect());
                self.import_item_completions( items, path.unwrap_or_default(), &source);
                return true;
            }
        }

        false
    }

    /// Add completions for all exports of a module.
    pub fn import_item_completions(
        &mut self,
        existing: ast::ImportItems,
        comps: Vec<ast::Ident>,
        source: &LinkedNode,
    ) {
        // Select the source by `comps`
        let value = self.worker.ctx.module_by_syntax(source);
        let value = comps
            .iter()
            .fold(value.as_ref(), |value, comp| value?.scope()?.get(comp));
        let Some(scope) = value.and_then(|v| v.scope()) else {
            return;
        };

        // Check imported items in the scope
        let seen = existing
            .iter()
            .flat_map(|item| {
                let item_comps = item.path().iter().collect::<Vec<_>>();
                if item_comps.len() == comps.len() + 1
                    && item_comps
                        .iter()
                        .zip(comps.as_slice())
                        .all(|(l, r)| l.as_str() == r.as_str())
                {
                    // item_comps.len() >= 1
                    item_comps.last().cloned()
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        if existing.iter().next().is_none() {
            self.snippet_completion("*", "*", "Import everything.");
        }

        for (name, value, _) in scope.iter() {
            if seen.iter().all(|item| item.as_str() != name) {
                self.value_completion(Some(name.clone()), value, false, None);
            }
        }
    }
}
