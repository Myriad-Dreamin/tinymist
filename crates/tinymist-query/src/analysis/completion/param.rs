//! Completion for param items.
//!
//! Note, this is used for the completion of parameters on a function's
//! *definition* instead of the completion of arguments of some *function call*.

use super::*;
impl CompletionPair<'_, '_, '_> {
    /// Complete parameters.
    pub fn complete_params(&mut self) -> bool {
        true
    }
}
