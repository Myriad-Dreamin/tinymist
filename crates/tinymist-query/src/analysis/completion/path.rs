use super::*;
impl CompletionPair<'_, '_, '_> {
    pub fn complete_path(&mut self, preference: &PathPreference) -> Option<Vec<CompletionItem>> {
        let id = self.cursor.source.id();
        if id.package().is_some() {
            return None;
        }

        let is_in_text;
        let text;
        let rng;
        // todo: the non-str case
        if self.cursor.leaf.is::<ast::Str>() {
            let vr = self.cursor.leaf.range();
            rng = vr.start + 1..vr.end - 1;
            if rng.start > rng.end
                || (self.cursor.cursor != rng.end && !rng.contains(&self.cursor.cursor))
            {
                return None;
            }

            let mut w = EcoString::new();
            w.push('"');
            w.push_str(&self.cursor.text[rng.start..self.cursor.cursor]);
            w.push('"');
            let partial_str = SyntaxNode::leaf(SyntaxKind::Str, w);

            text = partial_str.cast::<ast::Str>()?.get();
            is_in_text = true;
        } else {
            text = EcoString::default();
            rng = self.cursor.cursor..self.cursor.cursor;
            is_in_text = false;
        }
        crate::log_debug_ct!("complete_path: is_in_text: {is_in_text:?}");
        let path = Path::new(text.as_str());
        let has_root = path.has_root();

        let src_path = id.vpath();
        let base = id;
        let dst_path = src_path.join(path);
        let mut compl_path = dst_path.as_rootless_path();
        if !compl_path.is_dir() {
            compl_path = compl_path.parent().unwrap_or(Path::new(""));
        }
        crate::log_debug_ct!("compl_path: {src_path:?} + {path:?} -> {compl_path:?}");

        if compl_path.is_absolute() {
            log::warn!(
                "absolute path completion is not supported for security consideration {path:?}"
            );
            return None;
        }

        // find directory or files in the path
        let folder_completions = vec![];
        let mut module_completions = vec![];
        // todo: test it correctly
        for path in self.worker.ctx.completion_files(preference) {
            crate::log_debug_ct!("compl_check_path: {path:?}");

            // Skip self smartly
            if *path == base {
                continue;
            }

            let label: EcoString = if has_root {
                // diff with root
                unix_slash(path.vpath().as_rooted_path()).into()
            } else {
                let base = base
                    .vpath()
                    .as_rooted_path()
                    .parent()
                    .unwrap_or(Path::new("/"));
                let path = path.vpath().as_rooted_path();
                let w = pathdiff::diff_paths(path, base)?;
                unix_slash(&w).into()
            };
            crate::log_debug_ct!("compl_label: {label:?}");

            module_completions.push((label, CompletionKind::File));

            // todo: looks like the folder completion is broken
            // if path.is_dir() {
            //     folder_completions.push((label, CompletionKind::Folder));
            // }
        }

        let replace_range = self.cursor.lsp_range_of(rng);

        fn is_dot_or_slash(ch: &char) -> bool {
            matches!(*ch, '.' | '/')
        }

        let path_priority_cmp = |lhs: &str, rhs: &str| {
            // files are more important than dot started paths
            if lhs.starts_with('.') || rhs.starts_with('.') {
                // compare consecutive dots and slashes
                let a_prefix = lhs.chars().take_while(is_dot_or_slash).count();
                let b_prefix = rhs.chars().take_while(is_dot_or_slash).count();
                if a_prefix != b_prefix {
                    return a_prefix.cmp(&b_prefix);
                }
            }
            lhs.cmp(rhs)
        };

        module_completions.sort_by(|a, b| path_priority_cmp(&a.0, &b.0));
        // folder_completions.sort_by(|a, b| path_priority_cmp(&a.0, &b.0));

        let mut sorter = 0;
        let digits = (module_completions.len() + folder_completions.len())
            .to_string()
            .len();
        let completions = module_completions.into_iter().chain(folder_completions);
        Some(
            completions
                .map(|typst_completion| {
                    let lsp_snippet = &typst_completion.0;
                    let text_edit = EcoTextEdit::new(
                        replace_range,
                        if is_in_text {
                            lsp_snippet.clone()
                        } else {
                            eco_format!(r#""{lsp_snippet}""#)
                        },
                    );

                    let sort_text = eco_format!("{sorter:0>digits$}");
                    sorter += 1;

                    // todo: no all clients support label details
                    LspCompletion {
                        label: typst_completion.0,
                        kind: typst_completion.1,
                        detail: None,
                        text_edit: Some(text_edit),
                        // don't sort me
                        sort_text: Some(sort_text),
                        filter_text: Some("".into()),
                        insert_text_format: Some(InsertTextFormat::PLAIN_TEXT),
                        ..Default::default()
                    }
                })
                .collect_vec(),
        )
    }
}
