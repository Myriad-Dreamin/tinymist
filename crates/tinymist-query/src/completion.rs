use ecow::eco_format;
use lsp_types::{CompletionItem, CompletionList, CompletionTextEdit, InsertTextFormat, TextEdit};
use reflexo::path::{unix_slash, PathClean};

use crate::{
    prelude::*,
    syntax::{get_deref_target, DerefTarget},
    typst_to_lsp::completion_kind,
    upstream::{autocomplete, Completion, CompletionContext, CompletionKind},
    LspCompletion, StatefulRequest,
};

use self::typst_to_lsp::completion;

/// The [`textDocument/completion`] request is sent from the client to the
/// server to compute completion items at a given cursor position.
///
/// If computing full completion items is expensive, servers can additionally
/// provide a handler for the completion item resolve request
/// (`completionItem/resolve`). This request is sent when a completion item is
/// selected in the user interface.
///
/// [`textDocument/completion`]: https://microsoft.github.io/language-server-protocol/specification#textDocument_completion
///
/// # Compatibility
///
/// Since 3.16.0, the client can signal that it can resolve more properties
/// lazily. This is done using the `completion_item.resolve_support` client
/// capability which lists all properties that can be filled in during a
/// `completionItem/resolve` request.
///
/// All other properties (usually `sort_text`, `filter_text`, `insert_text`, and
/// `text_edit`) must be provided in the `textDocument/completion` response and
/// must not be changed during resolve.
#[derive(Debug, Clone)]
pub struct CompletionRequest {
    /// The path of the document to compute completions.
    pub path: PathBuf,
    /// The position in the document at which to compute completions.
    pub position: LspPosition,
    /// Whether the completion is triggered explicitly.
    pub explicit: bool,
}

impl StatefulRequest for CompletionRequest {
    type Response = CompletionResponse;

    fn request(
        self,
        ctx: &mut AnalysisContext,
        doc: Option<VersionedDocument>,
    ) -> Option<Self::Response> {
        let doc = doc.as_ref().map(|doc| doc.document.as_ref());
        let source = ctx.source_by_path(&self.path).ok()?;
        let cursor = {
            let mut cursor = ctx.to_typst_pos(self.position, &source)?;
            let text = source.text();

            // while is not char boundary, move cursor to right
            while cursor < text.len() && !text.is_char_boundary(cursor) {
                cursor += 1;
            }

            cursor
        };

        // Please see <https://github.com/nvarner/typst-lsp/commit/2d66f26fb96ceb8e485f492e5b81e9db25c3e8ec>
        //
        // FIXME: correctly identify a completion which is triggered
        // by explicit action, such as by pressing control and space
        // or something similar.
        //
        // See <https://github.com/microsoft/language-server-protocol/issues/1101>
        // > As of LSP 3.16, CompletionTriggerKind takes the value Invoked for
        // > both manually invoked (for ex: ctrl + space in VSCode) completions
        // > and always on (what the spec refers to as 24/7 completions).
        //
        // Hence, we cannot distinguish between the two cases. Conservatively, we
        // assume that the completion is not explicit.
        let explicit = false;

        let root = LinkedNode::new(source.root());
        let node = root.leaf_at(cursor);
        let deref_target = node.and_then(|node| get_deref_target(node, cursor));

        if let Some(d) = &deref_target {
            let node = d.node();
            // skip if is the let binding item, todo, check whether the pattern is exact
            if matches!(
                node.parent_kind(),
                Some(SyntaxKind::LetBinding | SyntaxKind::Closure)
            ) {
                return None;
            }
        }

        let mut match_ident = None;
        let mut completion_result = None;
        match deref_target {
            Some(DerefTarget::Callee(v) | DerefTarget::VarAccess(v)) => {
                if v.is::<ast::Ident>() {
                    match_ident = Some(v);
                }
            }
            Some(DerefTarget::ImportPath(v)) => {
                if !v.text().starts_with(r#""@"#) {
                    completion_result = complete_path(ctx, v, &source, cursor);
                }
            }
            None => {}
        }

        let items = completion_result.or_else(|| {
            let cc_ctx = CompletionContext::new(ctx, doc, &source, cursor, explicit)?;
            let (offset, mut completions) = autocomplete(cc_ctx)?;

            let replace_range;
            if match_ident.as_ref().is_some_and(|i| i.offset() == offset) {
                let match_ident = match_ident.unwrap();
                let rng = match_ident.range();
                replace_range = ctx.to_lsp_range(match_ident.range(), &source);

                let ident_prefix = source.text()[rng.start..cursor].to_string();
                completions.retain(|c| {
                    // c.label
                    let mut prefix_matcher = c.label.chars();
                    'ident_matching: for ch in ident_prefix.chars() {
                        for c in prefix_matcher.by_ref() {
                            if c == ch {
                                continue 'ident_matching;
                            }
                        }

                        return false;
                    }

                    true
                });
            } else {
                let lsp_start_position = ctx.to_lsp_pos(offset, &source);
                replace_range = LspRange::new(lsp_start_position, self.position);
            }

            Some(
                completions
                    .iter()
                    .map(|typst_completion| completion(typst_completion, replace_range))
                    .collect_vec(),
            )
        })?;

        // To response completions in fine-grained manner, we need to mark result as
        // incomplete. This follows what rust-analyzer does.
        // https://github.com/rust-lang/rust-analyzer/blob/f5a9250147f6569d8d89334dc9cca79c0322729f/crates/rust-analyzer/src/handlers/request.rs#L940C55-L940C75
        Some(CompletionResponse::List(CompletionList {
            is_incomplete: true,
            items,
        }))
    }
}

fn complete_path(
    ctx: &AnalysisContext,
    v: LinkedNode,
    source: &Source,
    cursor: usize,
) -> Option<Vec<CompletionItem>> {
    let id = source.id();
    if id.package().is_some() {
        return None;
    }

    let vp = v.cast::<ast::Str>()?;
    // todo: path escape
    let real_content = vp.get();
    let text = v.text();
    let unquoted = &text[1..text.len() - 1];
    if unquoted != real_content {
        return None;
    }

    let text = source.text();
    let vr = v.range();
    let offset = vr.start + 1;
    if cursor < offset || vr.end <= cursor || vr.len() < 2 {
        return None;
    }
    let path = Path::new(&text[offset..cursor]);
    let is_abs = path.is_absolute();

    let src_path = id.vpath();
    let base = src_path.resolve(&ctx.analysis.root)?;
    let dst_path = src_path.join(path);
    let mut compl_path = dst_path.as_rootless_path();
    if !compl_path.is_dir() {
        compl_path = compl_path.parent().unwrap_or(Path::new(""));
    }
    log::debug!("compl_path: {src_path:?} + {path:?} -> {compl_path:?}");

    if compl_path.is_absolute() {
        log::warn!("absolute path completion is not supported for security consideration {path:?}");
        return None;
    }

    let dirs = ctx.analysis.root.join(compl_path);
    log::debug!("compl_dirs: {dirs:?}");
    // find directory or files in the path
    let mut folder_completions = vec![];
    let mut module_completions = vec![];
    // todo: test it correctly
    for entry in dirs.read_dir().ok()? {
        let Ok(entry) = entry else {
            continue;
        };
        let path = entry.path();
        log::trace!("compl_check_path: {path:?}");
        if !path.is_dir() && !path.extension().is_some_and(|ext| ext == "typ") {
            continue;
        }
        if path.is_dir()
            && path
                .file_name()
                .is_some_and(|name| name.to_string_lossy().starts_with('.'))
        {
            continue;
        }

        // diff with root
        let path = dirs.join(path);

        // Skip self smartly
        if path.clean() == base.clean() {
            continue;
        }

        let label = if is_abs {
            // diff with root
            let w = path.strip_prefix(&ctx.analysis.root).ok()?;
            eco_format!("/{}", unix_slash(w))
        } else {
            let base = base.parent()?;
            let w = pathdiff::diff_paths(&path, base)?;
            unix_slash(&w).into()
        };
        log::debug!("compl_label: {label:?}");

        if path.is_dir() {
            folder_completions.push(Completion {
                label,
                kind: CompletionKind::Folder,
                apply: None,
                detail: None,
                command: None,
            });
        } else {
            module_completions.push(Completion {
                label,
                kind: CompletionKind::Module,
                apply: None,
                detail: None,
                command: None,
            });
        }
    }

    let rng = offset..vr.end - 1;
    let replace_range = ctx.to_lsp_range(rng, source);

    module_completions.sort_by(|a, b| a.label.cmp(&b.label));
    folder_completions.sort_by(|a, b| a.label.cmp(&b.label));

    let mut sorter = 0;
    let digits = (module_completions.len() + folder_completions.len())
        .to_string()
        .len();
    let completions = module_completions.into_iter().chain(folder_completions);
    Some(
        completions
            .map(|typst_completion| {
                let lsp_snippet = typst_completion
                    .apply
                    .as_ref()
                    .unwrap_or(&typst_completion.label);
                let text_edit =
                    CompletionTextEdit::Edit(TextEdit::new(replace_range, lsp_snippet.to_string()));

                let sort_text = format!("{sorter:0>digits$}");
                sorter += 1;

                let res = LspCompletion {
                    label: typst_completion.label.to_string(),
                    kind: Some(completion_kind(typst_completion.kind.clone())),
                    detail: typst_completion.detail.as_ref().map(String::from),
                    text_edit: Some(text_edit),
                    // don't sort me
                    sort_text: Some(sort_text),
                    filter_text: Some("".to_owned()),
                    insert_text_format: Some(InsertTextFormat::PLAIN_TEXT),
                    ..Default::default()
                };

                log::debug!("compl_res: {res:?}");

                res
            })
            .collect_vec(),
    )
}

#[cfg(test)]
mod tests {
    use insta::with_settings;
    use lsp_types::{CompletionItem, CompletionList};

    use super::*;
    use crate::tests::*;

    #[test]
    fn test() {
        snapshot_testing("completion", &|ctx, path| {
            let source = ctx.source_by_path(&path).unwrap();
            let rng = find_test_range(&source);
            let text = source.text()[rng.clone()].to_string();

            let mut results = vec![];
            for s in rng.clone() {
                let request = CompletionRequest {
                    path: path.clone(),
                    position: ctx.to_lsp_pos(s, &source),
                    explicit: false,
                };
                results.push(request.request(ctx, None).map(|resp| {
                    // CompletionResponse::Array(items)
                    match resp {
                        CompletionResponse::List(l) => CompletionResponse::List(CompletionList {
                            is_incomplete: l.is_incomplete,
                            items: l
                                .items
                                .into_iter()
                                .map(|item| CompletionItem {
                                    label: item.label,
                                    kind: item.kind,
                                    text_edit: item.text_edit,
                                    ..Default::default()
                                })
                                .collect(),
                        }),
                        CompletionResponse::Array(items) => CompletionResponse::Array(
                            items
                                .into_iter()
                                .map(|item| CompletionItem {
                                    label: item.label,
                                    kind: item.kind,
                                    text_edit: item.text_edit,
                                    ..Default::default()
                                })
                                .collect(),
                        ),
                    }
                }));
            }
            with_settings!({
                description => format!("Completion on {text} ({rng:?})"),
            }, {
                assert_snapshot!(JsonRepr::new_pure(results));
            })
        });
    }
}
