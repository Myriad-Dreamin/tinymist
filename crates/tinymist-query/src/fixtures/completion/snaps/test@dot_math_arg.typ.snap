---
source: crates/tinymist-query/src/completion.rs
description: Completion on / (31..32)
expression: "JsonRepr::new_pure(results)"
input_file: crates/tinymist-query/src/fixtures/completion/dot_math_arg.typ
---
[
 {
  "isIncomplete": false,
  "items": [
   {
    "kind": 15,
    "label": "align",
    "labelDetails": {
     "description": ".align"
    },
    "textEdit": {
     "newText": "align(${1:}, box(\\$1\\$))",
     "range": {
      "end": {
       "character": 10,
       "line": 2
      },
      "start": {
       "character": 10,
       "line": 2
      }
     }
    }
   }
  ]
 }
]
